subsys/bluetooth/host/aes_ccm.c

```diff
@@ -37,7 +37,7 @@ static inline void xor16(uint8_t *dst, const uint8_t *a, const uint8_t *b)
 
 /* pmsg is assumed to have the nonce already present in bytes 1-13 */
 static int ccm_calculate_X0(const uint8_t key[16], const uint8_t *aad, uint8_t aad_len,
-			    size_t mic_size, uint8_t msg_len, uint8_t b[16],
+			    size_t mic_size, uint16_t msg_len, uint8_t b[16],
```

Comment:
Why not just go with `size_t`, unless there's s a specific reason to stick to 16 bits?

Reply:
Yep. `msg_len` participates in authentication transformation. 
```
Form the 16-octet B0 field consisting of the 1-octet Flags field defined in step a) in this subclause,
the (15 – L)-octet Nonce field N, and the L-octet representation of the Length field l(m), as follows:
B0 = Flags || Nonce N || l(m).
```
l(m) is length of a plain message and has to be 2 bytes in case of Bluetooth.

Reply:
@alxelax Where is that text from? 

This is a quite hard function to understand due to missing source about the function. The function also references `pmsg` which is no where to be found, and I assume `pmsg` is the message of the length `msg_len`? 

The calling function, `ccm_auth` uses `size_t` for the `msg_len` with no value checks, so it's still possible that this is called with a value > 16-bit, so a value check and perhaps a cast to `uint16_t` might be useful there as well. 

`ccm_calculate_X0` also seems to only use `msg_len` to encode it into `b`, but why isn't the caller (`ccm_auth`) doing that? It Already encodes the first 14 octets of `b`. 

I'm sorry if you are unable to answer these questions and that's fine ;)

But at the very least, to fix this, we'll need to add the check that `msg_len` can fit in 16-bit, or change `ccm_auth` to also take a `uint16_t` (and the calling function of that will then also need to be updated. 

I see that the "top-level" functions that calls these do have checks for the length, so it should be trivial to change `msg_len` of `ccm_auth` to `uint16_t`, which I am in favor of.

Reply:
> @alxelax Where is that text from?
>
@Thalley, this is from ieee 802.15.4 standard. It uses the same encryption algorithms. Transformations are better described compared to the initial aes spec (at least from my point of view :) )
 
> This is a quite hard function to understand due to missing source about the function. The function also references `pmsg` which is no where to be found, and I assume `pmsg` is the message of the length `msg_len`?
> 

looking at log it seems the implementation was moved here `as is` from mesh. I guess mentioning of `pmsg` is leftover.

> The calling function, `ccm_auth` uses `size_t` for the `msg_len` with no value checks, so it's still possible that this is called with a value > 16-bit, so a value check and perhaps a cast to `uint16_t` might be useful there as well.
> 

yep, seems the potential way of doing bugs

> `ccm_calculate_X0` also seems to only use `msg_len` to encode it into `b`, but why isn't the caller (`ccm_auth`) doing that? It Already encodes the first 14 octets of `b`.
> 

no idea :)

> I'm sorry if you are unable to answer these questions and that's fine ;)
> 
> But at the very least, to fix this, we'll need to add the check that `msg_len` can fit in 16-bit, or change `ccm_auth` to also take a `uint16_t` (and the calling function of that will then also need to be updated.
> 
> I see that the "top-level" functions that calls these do have checks for the length, so it should be trivial to change `msg_len` of `ccm_auth` to `uint16_t`, which I am in favor of.

Ok